pkg/cli/app_metadata_migration.go

```diff
@@ -0,0 +1,68 @@
+package cli
+
+import (
+	"context"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/pyroscope-io/pyroscope/pkg/storage"
+	"github.com/sirupsen/logrus"
+)
+
+type AppNamesGetter interface {
+	GetAppNames(ctx context.Context) []string
+}
+
+type AppMetadataSaver interface {
+	CreateOrUpdate(_ context.Context, _ storage.ApplicationMetadata) error
+	List(context.Context) ([]storage.ApplicationMetadata, error)
+}
+
+type AppMetadataMigrator struct {
+	logger           *logrus.Logger
+	appNamesGetter   AppNamesGetter
+	appMetadataSaver AppMetadataSaver
+}
+
+func NewAppMetadataMigrator(logger *logrus.Logger, appNamesGetter AppNamesGetter, appMetadataSaver AppMetadataSaver) *AppMetadataMigrator {
+	return &AppMetadataMigrator{
+		logger:           logger,
+		appNamesGetter:   appNamesGetter,
+		appMetadataSaver: appMetadataSaver,
+	}
+}
+
+// Migrate creates Applications given a list of app names
+func (m *AppMetadataMigrator) Migrate() (err error) {
+	ctx := context.Background()
+
+	// Get all app names
+	appNamesFromOrigin := m.appNamesGetter.GetAppNames(ctx)
+	apps, err := m.appMetadataSaver.List(ctx)
+	if err != nil {
+		return err
+	}
+
+	// TODO skip if not necessary
+
+	// Convert slice -> map
+	appMap := make(map[string]storage.ApplicationMetadata)
+	for _, a := range apps {
+		appMap[a.FQName] = a
+	}
+
+	// If they don't exist already
+	for _, a := range appNamesFromOrigin {
+		if _, ok := appMap[a]; !ok {
+			logrus.Info("Migrating app: ", a)
+			// Write to MetadataSaver
+			saveErr := m.appMetadataSaver.CreateOrUpdate(ctx, storage.ApplicationMetadata{
+				FQName: a,
```

Comment:
does this mean that we won't populate any of the other metadata, e.g spy name, and it would only be added on the next ingest request?

and if so, how hard would it be to make it populate other fields as well?

Reply:
Requires querying latest profile, which I don't know a good way to do so, so I would query from EPOCH until now to be certain. I bet @kolesnikovae has a better idea.

But as I said previously, it's possible that there are applications without any other metadata (eg ones consumed via collapsed), so we have to deal with the case where they are absent anyway.

Reply:
I agree, it may be tricky. One approach is to pick the latest tree of that app (via reverse badger DB iterator with app name prefix) and then load the corresponding segment